### ISSUE TITLE
[panic] starred fixed-tuple base unpacked to a `@final` class panics `check_static_class_definitions`

### ISSUE THREAD

**Issue by kc0506** [astral-sh/ty#3293]
## Summary

While reviewing the fix in https://github.com/astral-sh/ruff/pull/24695 (which closed #3290), I noticed it only replaced `class_node.bases()[i]` indexing in the `try_mro` error-handling arms. The same bug pattern — using an *expanded*-bases index to index into the *AST* bases list — still exists in the [large per-base lint loop](https://github.com/astral-sh/ruff/blob/3834a4f0149dbb9c473331ed754d03b6c5bfcc5d/crates/ty_python_semantic/src/types/infer/builder/post_inference/static_class.rs#L221) in `check_static_class_definitions`. Several per-base diagnostics in that loop (`SUBCLASS_OF_FINAL_CLASS`, `INVALID_NAMED_TUPLE`, `INVALID_BASE`, `INVALID_PROTOCOL`, `INVALID_TYPED_DICT_HEADER`, `INVALID_GENERIC_CLASS`, frozen-dataclass inheritance) are likely to panic on out-of-bounds when a starred fixed-length-tuple base expands the semantic bases list past the AST bases list length.

I reproduce the simplest such case with `SUBCLASS_OF_FINAL_CLASS` below.

## Reproducer

```python
class X(*(int, bool)):
    pass
```

## Output (on `main` at 3834a4f)

```
error[panic]: Panicked at crates/ty_python_semantic/src/types/infer/builder/post_inference/static_class.rs:349:64 when checking `poc_final.py`: `index out of bounds: the len is 1 but the index is 1`
info: This indicates a bug in ty.
...
info: query stacktrace:
   0: infer_scope_types_impl(Id(1000))
             at crates/ty_python_semantic/src/types/infer.rs:263
   1: check_file_impl(Id(c00))
             at crates/ty_project/src/lib.rs:605
```

`static_class.rs:349` is the `SUBCLASS_OF_FINAL_CLASS` report inside the loop at line 221:

```rust
for (i, base_class) in class.explicit_bases(db).iter().enumerate() {
    // ...
    if base_class.is_final(db) {
        if let Some(builder) =
            context.report_lint(&SUBCLASS_OF_FINAL_CLASS, &class_node.bases()[i])
```

`class.explicit_bases(db)` contains the expanded list `[int, bool]` (len 2), but `class_node.bases()` is `[Starred(Tuple(int, bool))]` (len 1), so `class_node.bases()[1]` panics.

The same pattern produces additional panic paths in the same loop; `bool` just happens to be the easiest one to hit because it's `@final` in typeshed. Others cases could be constructed with starred tuples containing e.g. bare `Generic`, `Protocol[...]` subscripted bases mixed with PEP-695 type params, invalid TypedDict bases, etc, but I haven't tested them yet.

## Version

Tested on `main` at 3834a4f.

**Comment by charliermarsh** [comment]
Thanks!
